tech-specs/src/explore/design/dkg.md

```diff
@@ -1 +1,178 @@
+ - The DKG state machine associated to each of these session ids.
+
+##  The DKG state machine
+
+We first describe each new message type needed for the protocol.
+- `Deal`: This message contains the PVSS transcript from a validator
+  acting as a dealer
+- `Aggregate`: Add the combined PVSS transcripts as well as each transcript
+  included in the aggregation in the header of a proposed block
+- `Complain`: A message stating that a given validator's PVSS transcript or
+  a block proposer's aggregation failed verification. 
+
+In Ferveo, the DKG protocol is modelled as a state machine. This allows a
+validator to initialize an instance and drive it forward by reacting to
+each of the above message in the appropriate way. All necessary state
+is persisted by this state machine. We describe the  appropriate action for
+each of the above transactions.
+
+ - `Deal`: These need to be verified. Each validator adds every deal message
+   it receives to the state machine and when the threshold is reached,
+   the state machine automatically performs the aggregation. The subsequent
+   block proposer checks if the aggregation was performed. If so, it adds 
+   the result to the block it is proposing. In either case, it also adds the
+   PVSS transcripts to the block header. If aggregation was successful, these
+   act as verifications. Otherwise, the next block proposer will need the 
+   information to try and complete the  protocol in the next block. For any
+   PVSS transcript that fails verification, a `Complain` message is also 
+   included in the header.
+ - `Aggregate`: This simply needs to be verified and the resulting key added
+   to the DKG instance along with the input PVSS transcripts. If validation
+   of the aggregated key fails, a `Complain` should be issued.
+
+While every validator must send and react to `Deal` messages, `Aggregate` 
+must be made by the current block proposer if their current DKG state machine 
+is in the  appropriate state and/or the appropriate conditions are met. Thus 
+block proposers must check the state machines of their active DKG instances.
+
+## ABCI++
+
+The creation of the public key and the shared secrets requires ABCI++ as we
+need control over block preparation as well extra data from the Vote Extension
+phase. 
+
+We list the phase of ABCI++ in which each of the above actions takes place.
+
+ - `Deal`: `Deal` happens during the Vote Extension phase.
+ - `Aggregate`: This is done during the Prepare Proposal phase (it could be
+   done earlier if validators know they are likely to be the next proposer)
+   and will be  included in the header of the proposed block. The final result 
+   should be verified by validators. If validation fails, the next block 
+   proposer can try to aggregate the the PVSS transcripts correctly and 
+   include the result in the header of the subsequent block. Furthermore, 
+   `Complains` may be issued against the validator that incorrectly
```

Comment:
when does this happen w.r.t. ABCI++?

Reply:
Complaints got reworked a lot based on yesterday's conversation. They aren't an explicit thing anymore, so this is outdated.